std/assembly/util/string.ts

```diff
@@ -160,54 +167,211 @@ export function strtol<T>(str: string, radix: i32 = 0): T {
+    if (exp > 0) return significandf * pow10(exp);
+    return significandf / pow10(-exp);
+  } else if (exp < 0) {
+    return scaledown(significand, exp);
+  } else {
+    return scaleup(significand, exp);
+  }
+}
+
+// Adopted from metallic lib:
+// https://github.com/jdh8/metallic/blob/master/src/stdlib/parse/scientific.h
+// @ts-ignore: decorator
+@inline
+function scaledown(significand: u64, exp: i32): f64 {
+  const denom: u64 = 6103515625; // 1e14 * 0x1p-14
+  const scale = reinterpret<f64>(0x3F06849B86A12B9B); // 1e-14 * 0x1p32
+
+  var shift = clz(significand);
+  significand <<= shift;
+  shift = exp - shift;
+
+  for (; exp <= -14; exp += 14) {
+    let q = significand / denom;
+    let r = significand % denom;
+    let s = clz(q);
+    significand = (q << s) + <u64>nearest(scale * <f64>(r << (s - 18)));
+    shift -= s;
+  }
+  var b = <u64>ipow32(5, -exp);
+  var q = significand / b;
+  var r = significand % b;
+  var s = clz(q);
+  significand = (q << s) + <u64>(reinterpret<f64>(reinterpret<u64>(<f64>r) + (s << 52)) / <f64>b);
+  shift -= s;
+
+  return NativeMath.scalbn(<f64>significand, <i32>shift);
+}
+
+// Adopted from metallic lib:
+// https://github.com/jdh8/metallic/blob/master/src/stdlib/parse/scientific.h
+// @ts-ignore: decorator
+@inline
+function scaleup(significand: u64, exp: i32): f64 {
+  const coeff: u32 = 1220703125; // 1e13 * 0x1p-13;
+  var shift = ctz(significand);
+  significand >>= shift;
+  shift += exp;
+
+  __fixmulShift = shift;
+  for (; exp >= 13; exp -= 13) {
+    significand = fixmul(significand, coeff);
+  }
+  significand = fixmul(significand, <u32>ipow32(5, exp));
+  shift = __fixmulShift;
+  return NativeMath.scalbn(<f64>significand, <i32>shift);
+}
+
+// Adopted from metallic lib:
+// https://github.com/jdh8/metallic/blob/master/src/stdlib/parse/scientific.h
+// @ts-ignore: decorator
+@inline
+function parseExp(ptr: usize, len: i32): i32 {
+  var sign = 1, magnitude = 0;
+  var code = <u32>load<u16>(ptr);
+  // check code is 'e' or 'E'
+  if ((code | 32) != CharCode.e) return 0;
+
+  code = <u32>load<u16>(ptr += 2);
+  if (code == CharCode.MINUS) {
+    if (!--len) return 0;
+    code = <u32>load<u16>(ptr += 2);
+    sign = -1;
+  } else if (code == CharCode.PLUS) {
+    if (!--len) return 0;
+    code = <u32>load<u16>(ptr += 2);
+  }
+  // skip zeros
+  while (code == CharCode._0) {
+    if (!--len) return 0;
+    code = <u32>load<u16>(ptr += 2);
+  }
+  for (let digit: u32 = code - CharCode._0; len && digit < 10; digit = code - CharCode._0) {
+    if (magnitude >= 3200) return sign * 3200;
+    magnitude = 10 * magnitude + digit;
+    code = <u32>load<u16>(ptr += 2);
+    --len;
+  }
+  return sign * magnitude;
+}
+
+// @ts-ignore: decorator
+@lazy
+var __fixmulShift: u64 = 0;
+
+// Adopted from metallic lib:
+// https://github.com/jdh8/metallic/blob/master/src/stdlib/parse/scientific.h
+// @ts-ignore: decorator
+@inline
+function fixmul(a: u64, b: u32): u64 {
```

Comment:
This is called twice in scaleup. No strong opinion, but since inlined this would exist twice even if optimizing for size.

Reply:
Yes, I prefer to leave this inlined because it call in loop and in my experience this potentially could give significant speed boost especially because this touches global variable internally